pallets/crowdloan-rewards/src/lib.rs

```diff
@@ -51,124 +98,86 @@ pub mod pallet {
-			);
-			let current_reward_end_in = CurrentRewardEndIn::<T>::get();
-
-			ensure!(
-				&now >= &current_reward_end_in,
-				Error::<T>::AlreadyInitReward
-			);
-			for (account, amount) in &contributions {
-				let reward_info = RewardInfo {
-					total_reward: BalanceOf::<T>::from(*amount)
-						.saturating_mul(BalanceOf::<T>::from(rate)),
-					claimed_reward: 0u32.into(),
-					last_paid: now.clone(),
-				};
-				Contributors::<T>::insert(account, reward_info);
-			};
-			CurrentRewardEndIn::<T>::put(&end_block);
-			RewardPeriod::<T>::put(end_block - now);
-			Ok(Default::default())
-		}
-
-		#[pallet::weight(0)]
-		pub fn associate_account(
-			origin: OriginFor<T>,
-			relay_account: T::RelayChainAccountId,
-			proof: MultiSignature,
-		) -> DispatchResultWithPostInfo {
-			let who = ensure_signed(origin)?;
-			let payload = who.clone().encode();
-
-			ensure!(
-				proof.verify(payload.as_slice(), &relay_account.clone().into()),
-				Error::<T>::InvalidSignature
-			);
-
-			ensure!(
-				AssociatedAccount::<T>::get(&who).is_none(),
-				Error::<T>::AlreadyAssociated
-			);
-
-			AssociatedAccount::<T>::insert(&who, &relay_account);
-
-			Self::deposit_event(Event::AssociatedAccount(
-				who,
-				relay_account,
-			));
-			Ok(Default::default())
-		}
-
-		#[pallet::weight(0)]
-		pub fn update_associate_account(
-			origin: OriginFor<T>,
-			associated_native_account: T::AccountId,
-			relay_account: T::RelayChainAccountId,
-			proof: MultiSignature,
-		) -> DispatchResultWithPostInfo {
-			let who = ensure_signed(origin)?;
-			let payload = who.clone().encode();
 
-			ensure!(
-				proof.verify(payload.as_slice(), &relay_account.clone().into()),
-				Error::<T>::InvalidSignature
-			);
-			ensure!(
-				AssociatedAccount::<T>::get(&who).is_none(),
-				Error::<T>::AlreadyAssociated
-			);
-			let relay_associated_account = AssociatedAccount::<T>::get(&associated_native_account)
-				.ok_or(Error::<T>::NoAssociatedAccount)?;
-
-			ensure!(
-				relay_account == relay_associated_account,
-				Error::<T>::BadRelayAccount
-			);
-			AssociatedAccount::<T>::remove(associated_native_account);
-			AssociatedAccount::<T>::insert(&who, &relay_account);
-			Self::deposit_event(Event::UpdateAssociateAccount(
-				who,
-				relay_account,
-			));
+			let relay_block_now = BlockNumberOf::<T>::from(cumulus_pallet_parachain_system::Pallet::<T>::validation_data()
+				.expect("set_validation_data inherent needs to be present in every block!")
+				.relay_parent_number);
+			let setting = Settings::<T>::get();
+			ensure!(InitRewardAt::<T>::get().is_zero(), Error::<T>::AlreadyInitReward);
+
+			let total_reward_amount: BalanceOf<T> = contributions
+				.iter()
+				.fold(0u32.into(), |reward: BalanceOf<T>, (_, amount)| {
+				reward + *amount
+			});
+			ensure!(Self::pot() >= total_reward_amount, Error::<T>::InsufficientFunds);
+
+			for (who, amount) in &contributions {
+				let total_reward = BalanceOf::<T>::from(*amount);
```

Comment:
does is necessary? amount already typed as balance

Reply:
Removed it:
https://github.com/polkafoundry/newpolka/pull/55/commits/74449af91e637c9214423f0bf5f3f4697802ff21#diff-2bbae40c9c7fdfeeb0b1708d368844303a595083bc948ccd684b577e5d13f6ebR119

Reply:
@haicon2321993 don't need to de-ref

Reply:
Removed reference:
https://github.com/polkafoundry/newpolka/pull/55/commits/4e56dcf2ca96f49fa8399ada613475c71b6b783b#diff-2bbae40c9c7fdfeeb0b1708d368844303a595083bc948ccd684b577e5d13f6ebR119

Reply:
refactor code:
https://github.com/polkafoundry/newpolka/pull/55/commits/9913e86bea91dc1ec0dd74f14e52c5093dbd5b2a#diff-2bbae40c9c7fdfeeb0b1708d368844303a595083bc948ccd684b577e5d13f6ebR119